pkg/certificate/manager.go

```diff
@@ -1,59 +1,125 @@
 	"github.com/openservicemesh/osm/pkg/k8s/events"
 	"github.com/openservicemesh/osm/pkg/messaging"
 )
 
 // NewManager creates a new CertManager with the passed CA and CA Private Key
-func NewManager(mrcClient MRCClient, serviceCertValidityDuration time.Duration, msgBroker *messaging.Broker) (*Manager, error) {
-	// TODO(#4502): transition this call to a watch function that knows how to handle multiple MRC and can react to changes.
+func NewManager(ctx context.Context, mrcClient MRCClient, serviceCertValidityDuration time.Duration, msgBroker *messaging.Broker, checkInterval time.Duration) (*Manager, error) {
+	m := &Manager{
+		serviceCertValidityDuration: serviceCertValidityDuration,
+		msgBroker:                   msgBroker,
+	}
+	// We need an MRC in order to get the correct issuers, so handle initial MRC List now
 	mrcs, err := mrcClient.List()
 	if err != nil {
 		return nil, err
 	}
 
-	client, ca, clientID, err := mrcClient.GetCertIssuerForMRC(mrcs[0])
+	err = m.init(mrcs, mrcClient)
 	if err != nil {
 		return nil, err
 	}
 
-	c := &issuer{Issuer: client, ID: clientID, CertificateAuthority: ca}
-
-	m := &Manager{
-		// The signingIssuer is responsible for signing all newly issued certificates
-		// The validatingIssuer is the issuer that issued existing certificates.
-		// its underlying cert is still in the validating trust store
-		signingIssuer:               c,
-		validatingIssuer:            c,
-		serviceCertValidityDuration: serviceCertValidityDuration,
-		msgBroker:                   msgBroker,
+	// now that the manager is initialized, start a watch
+	// we wait until the manager is initialized so that there's
+	// no lock contention between the event handler and the initial
+	// state of the manager.
+	mrcEvents, err := mrcClient.Watch(ctx)
+	if err != nil {
+		return nil, err
 	}
-	return m, nil
-}
 
-// Start takes an interval to check if the certificate
-// needs to be rotated
-func (m *Manager) Start(checkInterval time.Duration, stop <-chan struct{}) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case event, open := <-mrcEvents:
+				if !open {
+					// channel was closed; return
+					return
+				}
+				m.handleMRCEvent(event)
+			}
+		}
+	}()
+
 	ticker := time.NewTicker(checkInterval)
 	go func() {
 		m.checkAndRotate()
 		for {
 			select {
-			case <-stop:
+			case <-ctx.Done():
 				ticker.Stop()
 				return
 			case <-ticker.C:
 				m.checkAndRotate()
 			}
 		}
 	}()
+
+	return m, nil
+}
+
+// init initializes the Manager receiver based on the observed state of the cluster at this
+// instant in time. This is why we pass in a slice of MRCs; we know our Watch call will make us initially
+// consistent, we just need a valid state of the world at this instant to kick things off.
+// This method is meant to be called at initialization time, so we don't grab the lock
+func (m *Manager) init(mrcs []*v1alpha2.MeshRootCertificate, mrcClient MRCClient) error {
```

Comment:
Is there a way we can get away with not doing this explicitly at startup and handle this just like an MRC added during runtime?

Then it seems like we could remove this constraint and let the informers do their regular list and watch without us having to filter out the initially-listed objects ourselves:
https://github.com/openservicemesh/osm/blob/f3aa02a10078d608c04e63dad20c23306e4c20d2/pkg/certificate/types.go#L129-L130

Reply:
So, the scenario I'm trying to avoid is when an Envoy proxy or the ADS server needs a certificate and there's no CA configured in the certificate manager at the moment. Especially since the `Start()` method was removed from `Manager`, the idea is that the `Manager` returned from `NewManager` is ready to use (or an error is returned). Yes that means we have to add some logic in the MRCClient, but I think the tradeoff is preferable

Reply:
I think this code might look almost identical to what we need for `handleMRCEvent` FWIW.

Might make sense to tackle in a single PR. thoughts @jaellio